src/lib/libast/man/aso.3
.fp 5 CW
.TH LIBASO 3
.SH NAME
\fBASO\fP \- Atomic Scalar Operations
.SH SYNOPSIS
.de Tp
.fl
.ne 2
.TP
..
.de Ss
.fl
.ne 2
.SS "\\$1"
..
.de Cs
.nf
.ft 5
..
.de Ce
.ft 1
.fi
..
.ta 1.0i 2.0i 3.0i 4.0i 5.0i
.Cs
#include <aso.h>
.Ce
.Ss "OPERATIONS"
.Cs
uint8_t		asocas8(uint8_t volatile*, int, int);
uint8_t		asoget8(uint8_t volatile*);
uint8_t		asoadd8(uint8_t volatile*, uint8_t);
uint8_t		asosub8(uint8_t volatile*, uint8_t);
uint8_t		asoinc8(uint8_t volatile*);
uint8_t		asodec8(uint8_t volatile*);

uint16_t	asocas16(uint16_t volatile*, uint16_t, uint16_t);
uint16_t	asoget16(uint16_t volatile*);
uint16_t	asoadd16(uint16_t volatile*, uint16_t);
uint16_t	asosub16(uint16_t volatile*, uint16_t);
uint16_t	asoinc16(uint16_t volatile*);
uint16_t	asodec16(uint16_t volatile*);

uint32_t	asocas32(uint32_t volatile*, uint32_t, uint32_t);
uint32_t	asoget32(uint32_t volatile*);
uint32_t	asoadd32(uint32_t volatile*, uint32_t);
uint32_t	asosub32(uint32_t volatile*, uint32_t);
uint32_t	asoinc32(uint32_t volatile*);
uint32_t	asodec32(uint32_t volatile*);

uint64_t	asocas64(uint64_t volatile*, uint64_t, uint64_t);
uint64_t	asoget64(uint64_t volatile*);
uint64_t	asoadd64(uint64_t volatile*, uint64_t);
uint64_t	asosub64(uint64_t volatile*, uint64_t);
uint64_t	asoinc64(uint64_t volatile*);
uint64_t	asodec64(uint64_t volatile*);

unsigned char	asocaschar(unsigned char volatile*, int, int);
unsigned char	asogetchar(unsigned char volatile*);
unsigned char	asoaddchar(unsigned char volatile*, int);
unsigned char	asosubchar(unsigned char volatile*, int);
unsigned char	asoincchar(unsigned char volatile*);
unsigned char	asodecchar(unsigned char volatile*);

unsigned short	asocasshort(unsigned short volatile*, unsigned short, unsigned short);
unsigned short	asogetshort(unsigned short volatile*);
unsigned short	asoaddshort(unsigned short volatile*, int);
unsigned short	asosubshort(unsigned short volatile*, int);
unsigned short	asoincshort(unsigned short volatile*);
unsigned short	asodecshort(unsigned short volatile*);

unsigned int	asocasint(unsigned int volatile*, unsigned int, unsigned int);
unsigned int	asogetint(unsigned int volatile*);
unsigned int	asoaddint(unsigned int volatile*, unsigned int);
unsigned int	asosubint(unsigned int volatile*, unsigned int);
unsigned int	asoincint(unsigned int volatile*);
unsigned int	asodecint(unsigned int volatile*);

unsigned long	asocaslong(unsigned long volatile*, unsigned long, unsigned long);
unsigned long	asogetlong(unsigned long volatile*);
unsigned long	asoaddlong(unsigned long volatile*, unsigned long);
unsigned long	asosublong(unsigned long volatile*, unsigned long);
unsigned long	asoinclong(unsigned long volatile*);
unsigned long	asodeclong(unsigned long volatile*);

size_t		asocassize(size_t volatile*, size_t, size_t);
size_t		asogetsize(size_t volatile*);
size_t		asoaddsize(size_t volatile*, size_t);
size_t		asosubsize(size_t volatile*, size_t);
size_t		asoincsize(size_t volatile*);
size_t		asodecsize(size_t volatile*);

void*		asocasptr(void volatile*, void*, void*);
void*		asogetptr(void volatile*);

int		asolock(unsigned int volatile*, unsigned int, int);

void		asospindecl(void);
void		asospininit(void);
void		asospinnext(void);
void		asospinrest(void);

#define		_ASO_INTRINSIC	[012]
.Ce
.SH DESCRIPTION
.PP
\fIASO\fP provides functions to perform atomic scalar operations.
The functions on the type \f5uint32_t\fP will be fully described below.
Other functions work similarly on their respective types.
64 bit operations are provided if the compiler supports 64 bit integers and/or pointers.

On most modern architectures (with modern compilers) the \fIASO\fP functions are macros
that reference the respective intrinsic operations directly.
Operations lacking intrinsics are implemented by composing other intrinsics.
At minimum a \fBcas\fP (compare and swap) atomic intrinsic must be available for
\fIASO\fP to ensure atomic operations.
The value of the macro \f5_ASO_INTRINSIC\fP describes the native environment:
.Tp
0
No atomic intrinsic operations available.
\fIASO\fP is emulated for a single process/thread and private memory only.
.Tp
1
Some atomic intrinsic operations are available.
\fIASO\fP works for multi process/thread and shared memory.
.Tp
2
Atomic intrinsics are implemented using atomic assembler machine instructions.
\fIASO\fP works for multi process/thread and shared memory.
.PP
.Ss "TYPES"
.PP
\f5uint8_t, uint16_t, uint32_t, uint64_t\fP

These are \fIunsigned integer\fP types of different sizes in bits.
For example, \f5uint32_t\fP represents the type of unsigned integer with 32 bits or 4 bytes.
.PP
.Ss "OPERATIONS"
.PP
.Ss "  uint32_t asoget32(uint32_t* from);"
This function returns the value \f5*from\fP.
.PP
.Ss "  uint32_t asoadd32(uint32_t* dest, uint32_t num);"
.Ss "  uint32_t asosub32(uint32_t* dest, uint32_t num);"
These functions add \f5num\fP to \f5*dest\fP and subtract \f5num\fP from \f5*dest\fP in an atomic step.
The return value is the old value in \f5*dest\fP.
.PP
.Ss "  uint32_t asoinc32(uint32_t* dest);"
.Ss "  uint32_t asodec32(uint32_t* dest);"
These functions increment \f5*dest\fP by 1 and decrement \f5*dest\fP by 1 in an atomic step.
The return value is the old value in \f5*dest\fP.

Consider an example where two concurrent threads/processes call \f5asoadd8()\fP
on the same \f5dest\fP with values, say \fIv1\fP and \fIv2\fP.
The eventual value in \f5dest\fP
will be as if \f5*dest += v1+v2\fP was performed in a single-threaded execution.

That should be contrasted with a situation where, instead of \f5asoadd8()\fP,
only normal arithmetics were used. Then, the end result could be any of
\f5*dest += v1\fP, \f5*dest += v2\fP or \f5*dest += v1+v2\fP, depending
on states of the hardware cache and the process scheduling.
.PP
.Ss "  uint32_t asocas32(uint32_t* dest, uint32_t tstval, uint32_t newval);"
This function provides the atomic \fIcompare-and-swap\fP operation.
If the current content of \f5dest\fP is equal to \f5tstval\fP then it will be set to \f5newval\fP.
If multiple threads/processes are performing the same operations only one will succeed with a
return value of \f5tstval\fP.
The return value is the old value in \f5*dest\fP.
.PP
.Ss "  int asolock(unsigned int* lock, unsigned int key, int type);"
This function uses \f5key\fP, a non-zero unsigned integer, to lock or unlock the \f5lock\fP.
It returns 0 or 1 on success and -1 on failure.

The argument \f5type\fP can be one of the below values:
.Tp
\f5ASO_TRYLOCK\fP:
This call makes a single attempt to use the given \f5key\fP to acquire a lock.
If successful, it returns 1; else it returns -1.
.Tp
\f5ASO_LOCK\fP:
This call uses a spin loop to acquire a lock with the given \f5key\fP.
On success, it returns 1. There is no failure mode with this call.
.Tp
\f5ASO_UNLOCK\fP:
This call unlocks the lock if it was locked with \f5key\fP. It returns 1 on success
and -1 on failure.
.PP
.Ss "  #define asospindecl()"
.Ss "  #define asospininit()"
.Ss "  #define asospinnext()"
.Ss "  #define asospinrest()"
These macro functions can be used to construct a concurrent spin-loop to process resources.
They attempt to balance out computing resource among threads and/or processes competing
for the same resources.
\f5asospindecl()\fP declares needed variables so it should be used in a syntactic context
suitable for such declarations.
\f5asospinrest()\fP sleeps for a fraction of a second.
Below is an example of how to manage a critical section of code using a spin-lock:

.Cs
      unsigned int lock;
      asospindecl();

      for(asospininit();; asospinnext() )
          if(asocasint(&lock, 0, 1) == 0)
               break;

      ... critical section of code ...

      asocasint(&lock, 1, 0); /* release lock */
.Ce

.PP
.SH AUTHOR
Kiem-Phong Vo, Adam Edgar, and Glenn Fowler


CHANGELOG.md
# ksh x.y.z (version TBD, this is a work in progress)

This is meant to document changes since the AST codoe (including the `ksh`
program) was open-sourced that will be in the next stable release based
on the original AST code. The next stable version will be treated as a
major release for several reasons. Not least of which is changing the
build tool chain from the legacy Nmake system to Meson. Legacy changes
can be found in the various `RELEASE` files.

Starting in June 2017 maintenance of the Korn shell (`ksh`) resumed with
the merging of some fixes from Red Hat by Siteshwar Vashisht after he was
granted commit privilege. In October 2017 Kurtis Rader noticed that `ksh`
had been open sourced and started contributing changes. This document
was subsequently created to document the work being done.

## Deprecations

None at this time.

## Notable non-backward compatible changes

- Support for the `universe` command has been removed (issue #793).
- Support for building on systems using EBCDIC has been removed (issue #742).
- Support for the `LC_OPTIONS` env var has been removed (issue #579).
- `case "[0-9]" in [0-9]) echo match;; esac` has stopped matching. When a case
  statement doesn't match a pattern, it no longer tries to use the pattern as
  a literal string (issue #476).
- echo builtin now interprets escape sequences and parses command line options
  on all platforms. (issue #370)
- Support for the UWIN environment has been removed (issue #284).
- The experimental `SHOPT_FIXEDARRAY` feature has been removed. It is known to
  have been enabled in some Fedora and RHEL builds. But because it was
  experimental and undocumented this should not break any existing users of
  `ksh` built with this feature enabled. (issue #234)
- <> operator now redirects standard input by default (issue #75).
- Support for the build time SHOPT_ACCTFILE symbol and code has been removed
  (issue #210).

## Notable fixes and improvements

- The project now passes its unit tests when built with malloc debugging
  enabled (i.e., `meson test --setup=malloc`).
- Changes to the project are now validated by running unit tests on the Travis
  continuous integration system.
- The ksh source now builds on BSD based systems such as macOS and FreeBSD.
- The ksh source now builds on Cygwin; albeit with many unit test failures.
- The legacy Nmake build system has been replaced by Meson. This improves the
  build time by roughly an order of magnitude (issue #42).

## Other significant changes

- The AST locale subsystem has been replaced by the platform's implementation
  (issue #579).
- The AST Vmalloc subsystem has been removed. The project now uses the system
  malloc. (issue #396)
- Operations that depend on the simulated 3DFS behavior found in release
  ksh93v are no longer supported. That behavior does not work in ksh93u (the
  most recent stable release found in all distros) unless the 3D file system
  is actually present. This includes operations such as this:
  `exec {fd}</dev; cd ~{fd}` (issue #510).
- The math `scalb` function has been removed since it was marked obsolete by
  POSIX in 2001 and is no longer found in some libc implementations.
- Code hidden behind the SHOPT_ACCT and SHOPT_ACCTFILE build time symbols
  has been removed. The features are not enabled by default and I am not aware
  of any distro which enables them. Furthermore, they are huge security holes
  and should never be enabled.
- The suid_exec program has been removed (issue #366).
- The code has been restyled according to new guidelines (issue #125).
- Many features which used to be optionally included at build time are now
  unconditionally included (e.g., the code protected by `SHOPT_MULTIBYTE`,
  `SHOPT_COMPLETE`, `SHOPT_BRACEPAT`, `SHOPT_RAWONLY`, `SHOPT_STATS,
  `SHOPT_OPTIMIZE`, `SHOPT_SUID_EXEC`, `SHOPT_FILESCAN`, `SHOPT_POLL,
  `SHOPT_AUDIT`, and `SHOPT_SYSRC`).
- Unit tests can now be run under `valgrind` to help detect more bugs.
- Any code not needed to build and run `ksh` has been removed from the master
  branch.
- Fixes backported from OpenSuse (issue #377).
- Fixes backported from Red Hat Fedora and Enterprise Linux (RHEL) distros
  (issue #172).
- Fixes backported from Solaris (issue #122).


src/lib/libast/aso/aso.c
/***********************************************************************
 *                                                                      *
 *               This software is part of the ast package               *
 *          Copyright (c) 1985-2013 AT&T Intellectual Property          *
 *                      and is licensed under the                       *
 *                 Eclipse Public License, Version 1.0                  *
 *                    by AT&T Intellectual Property                     *
 *                                                                      *
 *                A copy of the License is available at                 *
 *          http://www.eclipse.org/org/documents/epl-v10.html           *
 *         (with md5 checksum b35adb5213ca9657e911e9befb180842)         *
 *                                                                      *
 *              Information and Software Systems Research               *
 *                            AT&T Research                             *
 *                           Florham Park NJ                            *
 *                                                                      *
 *               Glenn Fowler <glenn.s.fowler@gmail.com>                *
 *                    David Korn <dgkorn@gmail.com>                     *
 *                     Phong Vo <phongvo@gmail.com>                     *
 *                                                                      *
 ***********************************************************************/
#include "config_ast.h"  // IWYU pragma: keep

#include <stdint.h>
#include <sys/types.h>

#include "aso.h"
#include "ast_aso.h"

/*
 * ast atomic scalar operations
 * AT&T Research
 *
 * cas { 8 16 32 [64] } subset snarfed from the work by
 * Adam Edgar and Kiem-Phong Vo 2010-10-10
 *
 * additional intrinsics and emulations by
 * Glenn Fowler 2011-11-11
 */

typedef union {
    uint8_t c[2];
    uint16_t i;
} U16_8_t;

typedef union {
    uint8_t c[4];
    uint32_t i;
} U32_8_t;

typedef union {
    uint16_t c[2];
    uint32_t i;
} U32_16_t;

#ifdef _ast_int8_t

typedef union {
    uint8_t c[8];
    uint64_t i;
} U64_8_t;

typedef union {
    uint16_t c[4];
    uint64_t i;
} U64_16_t;

typedef union {
    uint32_t c[2];
    uint64_t i;
} U64_32_t;

#if !_ASO_INTRINSIC || defined(_ast_int8_t) && !defined(asocas64)

/*
 * used only when no cas intrinsic is available
 * should work most of the time in a single threaded process
 * with no multi-process shared mem access and low freq signals
 */

static ssize_t lock(ssize_t k) {
    static int locked;

    if (k >= 0) {
        locked--;
        return 0;
    }
    while (locked++) locked--;
    return 1;
}

#endif

#if _ASO_Interlocked

#if (_BLD_posix || __CYGWIN__) && !_X64

#if __CYGWIN__

#include <dlfcn.h>

#define MODULE_kernel 0
#define getsymbol(m, s) dlsym(m, s)

#else

#include "dl.h"

#endif

typedef struct LL_s {
    LONG a;
    LONG b;
} LL_t;

typedef union {
    LONGLONG i;
    LL_t ll;
} LL_u;

static LONGLONG _aso_InterlockedCompareExchange64_init(LONGLONG volatile *, LONGLONG, LONGLONG);

_aso_InterlockedCompareExchange64_f _aso_InterlockedCompareExchange64 =
    _aso_InterlockedCompareExchange64_init;

static LONGLONG _aso_InterlockedCompareExchange64_32(LONGLONG volatile *p, LONGLONG o, LONGLONG n) {
    LL_t *lp = (LL_t *)p;
    LL_t *op = (LL_t *)&o;
    LL_t *np = (LL_t *)&n;
    LONGLONG r;

    r = *p;
    if (asocas32(&lp->a, op->a, np->a) == op->a) {
        if (asocas32(&lp->b, op->b, np->b) == op->b) return o;
        asocas32(&lp->a, np->a, op->a);
    }
    return r;
}

static LONGLONG _aso_InterlockedCompareExchange64_init(LONGLONG volatile *p, LONGLONG o,
                                                       LONGLONG n) {
    if (!(_aso_InterlockedCompareExchange64 = (_aso_InterlockedCompareExchange64_f)getsymbol(
              MODULE_kernel, "InterlockedCompareExchange64")))
        _aso_InterlockedCompareExchange64 = _aso_InterlockedCompareExchange64_32;
    return _aso_InterlockedCompareExchange64(p, o, n);
}

static LONGLONG _aso_InterlockedExchangeAdd64_init(LONGLONG volatile *, LONGLONG);

_aso_InterlockedExchangeAdd64_f _aso_InterlockedExchangeAdd64 = _aso_InterlockedExchangeAdd64_init;

static LONGLONG _aso_InterlockedExchangeAdd64_32(LONGLONG volatile *p, LONGLONG n) {
    LONGLONG o;

    do {
        o = *p;
    } while (_aso_InterlockedCompareExchange64_32(p, o, o + n) != o);
    return o;
}

static LONGLONG _aso_InterlockedExchangeAdd64_init(LONGLONG volatile *p, LONGLONG n) {
    if (!(_aso_InterlockedExchangeAdd64 = (_aso_InterlockedExchangeAdd64_f)getsymbol(
              MODULE_kernel, "InterlockedExchangeAdd64")))
        _aso_InterlockedExchangeAdd64 = _aso_InterlockedExchangeAdd64_32;
    return _aso_InterlockedExchangeAdd64(p, n);
}

#endif

#elif _ASO_i386

uint32_t _aso_cas32(uint32_t volatile *p, uint32_t o, uint32_t n) {
    uint32_t r;

    __asm__ __volatile__("lock ; cmpxchg %3,%4"
                         : "=a"(r), "=m"(*p)
                         : "0"(o), "q"(n), "m"(*p)
                         : "memory", "cc");
    return r;
}

#if _ast_sizeof_pointer == 8

uint64_t _aso_cas64(uint64_t volatile *p, uint64_t o, uint64_t n) {
    uint64_t r;

    __asm__ __volatile__("lock ; cmpxchg %3,%4"
                         : "=a"(r), "=m"(*p)
                         : "0"(o), "q"(n), "m"(*p)
                         : "memory", "cc");
    return r;
}

#endif

#elif _ASO_ia64

uint32_t _aso_cas32(uint32_t volatile *p, uint32_t o, uint32_t n) {
    int r;

    __asm__ __volatile__(
        "0:	lwarx %0,0,%1 ;"
        "	xor. %0,%3,%0;"
        "	bne 1f;"
        "	stwcx. %2,0,%1;"
        "	bne- 0b;"
        "1:"
        : "=&r"(r)
        : "r"(p), "r"(n), "r"(o)
        : "cr0", "memory");
    __asm__ __volatile__("isync" : : : "memory");
    return r ? *p : o;
}

uint64_t _aso_cas64(uint64_t volatile *p, uint64_t o, uint64_t n) {
    long r;

    __asm__ __volatile__(
        "0:	ldarx %0,0,%1 ;"
        "	xor. %0,%3,%0;"
        "	bne 1f;"
        "	stdcx. %2,0,%1;"
        "	bne- 0b;"
        "1:"
        : "=&r"(r)
        : "r"(p), "r"(n), "r"(o)
        : "cr0", "memory");
    __asm__ __volatile__("isync" : : : "memory");
    return r ? *p : o;
}

#elif _ASO_ppc

uint32_t _aso_cas32(uint32_t volatile *p, uint32_t o, uint32_t n) {
    int r;

    __asm__ __volatile__(
        "0:	lwarx %0,0,%1 ;"
        "	xor. %0,%3,%0;"
        "	bne 1f;"
        "	stwcx. %2,0,%1;"
        "	bne- 0b;"
        "1:"
        : "=&r"(r)
        : "r"(p), "r"(n), "r"(o)
        : "cr0", "memory");
    __asm__ __volatile__("isync" : : : "memory");
    return r ? *p : o;
}

uint64_t _aso_cas64(uint64_t volatile *p, uint64_t o, uint64_t n) {
    long r;

    __asm__ __volatile__(
        "0:	ldarx %0,0,%1 ;"
        "	xor. %0,%3,%0;"
        "	bne 1f;"
        "	stdcx. %2,0,%1;"
        "	bne- 0b;"
        "1:"
        : "=&r"(r)
        : "r"(p), "r"(n), "r"(o)
        : "cr0", "memory");
    __asm__ __volatile__("isync" : : : "memory");
    return r ? *p : o;
}

#endif

/*
 * sync and return "current" value
 */

#ifndef asoget8
uint8_t asoget8(uint8_t volatile *p) {
    int o;

    do {
        o = *p;
    } while (asocas8(p, o, o) != o);
    return o;
}
#endif

#ifndef asoget16
uint16_t asoget16(uint16_t volatile *p) {
    int o;

    do {
        o = *p;
    } while (asocas16(p, o, o) != o);
    return o;
}
#endif

#ifndef asoget32
uint32_t asoget32(uint32_t volatile *p) {
    uint32_t o;

    do {
        o = *p;
    } while (asocas32(p, o, o) != o);
    return o;
}
#endif

#ifdef _ast_int8_t

#ifndef asoget64
uint64_t asoget64(uint64_t volatile *p) {
    uint64_t o;

    do {
        o = *p;
    } while (asocas64(p, o, o) != o);
    return o;
}
#endif

#endif

#ifndef asogetptr
void *asogetptr(void volatile *p) {
    void *o;

    do {
        o = *(void *volatile *)p;
    } while (asocasptr((void **)p, o, o) != o);
    return o;
}
#endif

/*
 * add and return old value
 */

#ifndef asoadd8
uint8_t asoadd8(uint8_t volatile *p, int n) {
    int o;

    do {
        o = *p;
    } while (asocas8(p, o, o + n) != o);
    return o;
}
#endif

#ifndef asoadd16
uint16_t asoadd16(uint16_t volatile *p, int n) {
    int o;

    do {
        o = *p;
    } while (asocas16(p, o, o + n) != o);
    return o;
}
#endif

#ifndef asoadd32
uint32_t asoadd32(uint32_t volatile *p, uint32_t n) {
    uint32_t o;

    do {
        o = *p;
    } while (asocas32(p, o, o + n) != o);
    return o;
}
#endif

#ifdef _ast_int8_t

#ifndef asoadd64
uint64_t asoadd64(uint64_t volatile *p, uint64_t n) {
    uint64_t o;

    do {
        o = *p;
    } while (asocas64(p, o, o + n) != o);
    return o;
}
#endif

#endif

/*
 * subtract and return old value
 */

#ifndef asosub8
uint8_t asosub8(uint8_t volatile *p, int n) {
    int o;

    do {
        o = *p;
    } while (asocas8(p, o, o - n) != o);
    return o;
}
#endif

#ifndef asosub16
uint16_t asosub16(uint16_t volatile *p, int n) {
    int o;

    do {
        o = *p;
    } while (asocas16(p, o, o - n) != o);
    return o;
}
#endif

#ifndef asosub32
uint32_t asosub32(uint32_t volatile *p, uint32_t n) {
    uint32_t o;

    do {
        o = *p;
    } while (asocas32(p, o, o - n) != o);
    return o;
}
#endif

#ifdef _ast_int8_t

#ifndef asosub64
uint64_t asosub64(uint64_t volatile *p, uint64_t n) {
    uint64_t o;

    do {
        o = *p;
    } while (asocas64(p, o, o - n) != o);
    return o;
}
#endif

#endif

/*
 * increment and return old value
 */

#ifndef asoinc8
uint8_t asoinc8(uint8_t volatile *p) {
    int o;

    do {
        o = *p;
    } while (asocas8(p, o, o + 1) != o);
    return o;
}
#endif

#ifndef asoinc16
uint16_t asoinc16(uint16_t volatile *p) {
    int o;

    do {
        o = *p;
    } while (asocas16(p, o, o + 1) != o);
    return o;
}
#endif

#ifndef asoinc32
uint32_t asoinc32(uint32_t volatile *p) {
    uint32_t o;

    do {
        o = *p;
    } while (asocas32(p, o, o + 1) != o);
    return o;
}
#endif

#ifdef _ast_int8_t

#ifndef asoinc64
uint64_t asoinc64(uint64_t volatile *p) {
    uint64_t o;

    do {
        o = *p;
    } while (asocas64(p, o, o + 1) != o);
    return o;
}
#endif

#endif

/*
 * decrement and return old value
 */

#ifndef asodec8
uint8_t asodec8(uint8_t volatile *p) {
    int o;

    do {
        o = *p;
    } while (asocas8(p, o, o - 1) != o);
    return o;
}
#endif

#ifndef asodec16
uint16_t asodec16(uint16_t volatile *p) {
    int o;

    do {
        o = *p;
    } while (asocas16(p, o, o - 1) != o);
    return o;
}
#endif

#ifndef asodec32
uint32_t asodec32(uint32_t volatile *p) {
    uint32_t o;

    do {
        o = *p;
    } while (asocas32(p, o, o - 1) != o);
    return o;
}
#endif

#ifdef _ast_int8_t

#ifndef asodec64
uint64_t asodec64(uint64_t volatile *p) {
    uint64_t o;

    do {
        o = *p;
    } while (asocas64(p, o, o - 1) != o);
    return o;
}
#endif

#endif

/*
 * if *p <= n then return *p
 * else *p = n and return n
 */

#ifndef asomin8
uint8_t asomin8(uint8_t volatile *p, int n) {
    int o;

    for (;; asospinrest()) {
        if ((o = *p) <= n) return o;
        if (asocas8(p, o, n) == o) break;
    }
    return n;
}
#endif

#ifndef asomin16
uint16_t asomin16(uint16_t volatile *p, int n) {
    int o;

    for (;; asospinrest()) {
        if ((o = *p) <= n) return o;
        if (asocas16(p, o, n) == o) break;
    }
    return o;
}
#endif

#ifndef asomin32
uint32_t asomin32(uint32_t volatile *p, uint32_t n) {
    uint32_t o;

    for (;; asospinrest()) {
        if ((o = *p) <= n) return o;
        if (asocas32(p, o, n) == o) break;
    }
    return o;
}
#endif

#ifdef _ast_int8_t

#ifndef asomin64
uint64_t asomin64(uint64_t volatile *p, uint64_t n) {
    uint64_t o;

    for (;; asospinrest()) {
        if ((o = *p) <= n) return o;
        if (asocas64(p, o, n) == o) break;
    }
    return o;
}
#endif

#endif

/*
 * if *p >= n then return *p
 * else *p = n and return n
 */

#ifndef asomax8
uint8_t asomax8(uint8_t volatile *p, int n) {
    int o;

    for (;; asospinrest()) {
        if ((o = *p) >= n) return o;
        if (asocas8(p, o, n) == o) break;
    }
    return n;
}
#endif

#ifndef asomax16
uint16_t asomax16(uint16_t volatile *p, int n) {
    int o;

    for (;; asospinrest()) {
        if ((o = *p) >= n) return o;
        if (asocas16(p, o, n) == o) break;
    }
    return o;
}
#endif

#ifndef asomax32
uint32_t asomax32(uint32_t volatile *p, uint32_t n) {
    uint32_t o;

    for (;; asospinrest()) {
        if ((o = *p) >= n) return o;
        if (asocas32(p, o, n) == o) break;
    }
    return o;
}
#endif

#ifdef _ast_int8_t

#ifndef asomax64
uint64_t asomax64(uint64_t volatile *p, uint64_t n) {
    uint64_t o;

    for (;; asospinrest()) {
        if ((o = *p) >= n) return o;
        if (asocas64(p, o, n) == o) break;
    }
    return o;
}
#endif

#endif

/*
 * { 8 16 32 [64] } compare with old, swap with new if same, and return old value
 */

#ifndef asocas8
uint8_t asocas8(uint8_t volatile *p, int o, int n) {
#if defined(asocas16)
    U16_8_t u;
    U16_8_t v;
    U16_8_t *a;
    int s;
    int i;

    s = (int)(integralof(p) & (sizeof(u.i) - 1));
    a = (U16_8_t *)((char *)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
    for (;;) {
        u.i = a->i;
        u.c[s] = o;
        v.i = u.i;
        v.c[s] = n;
        if (asocas16(&a->i, u.i, v.i) == u.i) break;
        for (i = 0;; i++)
            if (i >= elementsof(u.c))
                return a->c[s];
            else if (i != s && u.c[i] != a->c[i])
                break;
    }
    return o;
#elif defined(asocas32)
    U32_8_t u;
    U32_8_t v;
    U32_8_t *a;
    int s;
    int i;

    s = (int)(integralof(p) & (sizeof(u.i) - 1));
    a = (U32_8_t *)((char *)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
    for (;;) {
        u.i = a->i;
        u.c[s] = o;
        v.i = u.i;
        v.c[s] = n;
        if (asocas32(&a->i, u.i, v.i) == u.i) break;
        for (i = 0;; i++)
            if (i >= elementsof(u.c))
                return a->c[s];
            else if (i != s && u.c[i] != a->c[i])
                break;
    }
    return o;
#elif defined(asocas64)
    U64_8_t u;
    U64_8_t v;
    U64_8_t *a;
    int s;
    int i;

    s = (int)(integralof(p) & (sizeof(u.i) - 1));
    a = (U64_8_t *)((char *)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
    for (;;) {
        u.i = a->i;
        u.c[s] = o;
        v.i = u.i;
        v.c[s] = n;
        if (asocas64(&a->i, u.i, v.i) == u.i) break;
        for (i = 0;; i++)
            if (i >= elementsof(u.c))
                return a->c[s];
            else if (i != s && u.c[i] != a->c[i])
                break;
    }
    return o;
#else
    ssize_t k;

    k = lock(0);
    if (*p == o)
        *p = n;
    else
        o = *p;
    lock(k);
    return o;
#endif
}
#endif

#ifndef asocas16
uint16_t asocas16(uint16_t volatile *p, int o, int n) {
#if defined(asocas32)
    U32_8_t u;
    U32_8_t v;
    U32_8_t *a;
    int s;
    int i;

    s = (int)(integralof(p) & (sizeof(u.i) - 1));
    a = (U32_8_t *)((char *)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
    for (;;) {
        u.i = a->i;
        u.c[s] = o;
        v.i = u.i;
        v.c[s] = n;
        if (asocas32(&a->i, u.i, v.i) == u.i) break;
        for (i = 0;; i++)
            if (i >= elementsof(u.c))
                return a->c[s];
            else if (i != s && u.c[i] != a->c[i])
                break;
    }
    return o;
#elif defined(asocas64)
    U64_8_t u;
    U64_8_t v;
    U64_8_t *a;
    int s;
    int i;

    s = (int)(integralof(p) & (sizeof(u.i) - 1));
    a = (U64_8_t *)((char *)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
    for (;;) {
        u.i = a->i;
        u.c[s] = o;
        v.i = u.i;
        v.c[s] = n;
        if (asocas64(&a->i, u.i, v.i) == u.i) break;
        for (i = 0;; i++)
            if (i >= elementsof(u.c))
                return a->c[s];
            else if (i != s && u.c[i] != a->c[i])
                break;
    }
    return o;
#else
    ssize_t k;

    k = lock(0);
    if (*p == o)
        *p = n;
    else
        o = *p;
    lock(k);
    return o;
#endif
}
#endif

#ifndef asocas32
uint32_t asocas32(uint32_t volatile *p, uint32_t o, uint32_t n) {
#if defined(asocas64)
    U64_8_t u;
    U64_8_t v;
    U64_8_t *a;
    int s;
    int i;

    s = (int)(integralof(p) & (sizeof(u.i) - 1));
    a = (U64_8_t *)((char *)0 + (integralof(p) & ~(sizeof(u.i) - 1)));
    for (;;) {
        u.i = a->i;
        u.c[s] = o;
        v.i = u.i;
        v.c[s] = n;
        if (asocas64(&a->i, u.i, v.i) == u.i) break;
        for (i = 0;; i++)
            if (i >= elementsof(u.c))
                return a->c[s];
            else if (i != s && u.c[i] != a->c[i])
                break;
    }
    return o;
#else
    ssize_t k;

    k = lock(0);
    if (*p == o)
        *p = n;
    else
        o = *p;
    lock(k);
    return o;
#endif
}
#endif

#ifdef _ast_int8_t

#ifndef asocas64
uint64_t asocas64(uint64_t volatile *p, uint64_t o, uint64_t n) {
    ssize_t k;

    k = lock(0);
    if (*p == o)
        *p = n;
    else
        o = *p;
    lock(k);
    return o;
}
#endif

#endif

/*
 * compare with old, swap with new if same, and return old value
 */

#ifndef asocasptr
void *asocasptr(void volatile *p, void *o, void *n) {
    ssize_t k;

    k = lock(0);
    if (*(void *volatile *)p == o)
        *(void *volatile *)p = n;
    else
        o = *(void *volatile *)p;
    lock(k);
    return o;
}
#endif  // asocasptr

#endif  // _ast_int8_t
